File: app/src/main/java/com/fsoft/vktest/ViewsLayer/MessagesFragment/MessageHistory.java
package com.fsoft.vktest.ViewsLayer.MessagesFragment;

import com.fsoft.vktest.AnswerInfrastructure.BotBrain;
import com.fsoft.vktest.AnswerInfrastructure.Message;
import com.fsoft.vktest.ApplicationManager;

import java.util.ArrayList;
import java.util.Objects;

/*
* Этот класс занимается хранением истории сообщений для отображения её на экране "сообщения".
* Вызывать из ApplicationManager!
* Обращаться  из MessagesFragment через ApplicationManager
*
* */
public class MessageHistory {
    private static int MESSAGE_LIMIT = 500;
    private ArrayList<MessageStatus> messages = new ArrayList<>(); //сначала самые новые, 0=последнее
    private ApplicationManager applicationManager = null;


    public MessageHistory(ApplicationManager applicationManager) {
        this.applicationManager = applicationManager;
        applicationManager.getBrain().addMessageListener(new BotBrain.OnMessageStatusChangedListener() {
            @Override
            public void messageReceived(Message message) {
                messages.add(0, new MessageStatus(message).received());
                //почистить очередь больше 100
                while(messages.size() > MESSAGE_LIMIT)
                    messages.remove(MESSAGE_LIMIT);
            }

            @Override
            public void messageAnswered(Message message) {
                for (MessageStatus messageStatus : messages)
                    if(messageStatus.getMessage().getMessage_id() == message.getMessage_id())
                        messageStatus.answered().setMessage(message);

            }

            @Override
            public void messageError(Message message, Exception e) {
                for (MessageStatus messageStatus : messages)
                    if(messageStatus.getMessage().getMessage_id() == message.getMessage_id())
                        messageStatus.error().setMessage(message);
            }

            @Override
            public void messageIgnored(Message message) {
                for (MessageStatus messageStatus : messages)
                    if(messageStatus.getMessage().getMessage_id() == message.getMessage_id())
                        messageStatus.ignored().setMessage(message);
            }
        });
    }
    public ArrayList<MessageStatus> getMessages() {
        return messages;
    }
}

File: app/src/main/java/com/fsoft/vktest/NewViewsLayer/MessagesFragment/MessageHistory.kt
package com.fsoft.vktest.NewViewsLayer.MessagesFragment

import com.fsoft.vktest.AnswerInfrastructure.BotBrain.OnMessageStatusChangedListener
import com.fsoft.vktest.AnswerInfrastructure.Message
import com.fsoft.vktest.ApplicationManager


/*
* Этот класс занимается хранением истории сообщений для отображения её на экране "сообщения".
* Вызывать из ApplicationManager!
* Обращаться  из MessagesFragment через ApplicationManager
*
* */
class MessageHistory(applicationManager: ApplicationManager) {
    val messages: MutableList<MessageStatus> = mutableListOf() //сначала самые новые, 0=последнее
//    private var applicationManager: ApplicationManager? = null


    init {
//        this.applicationManager = applicationManager
        applicationManager.brain.addMessageListener(object : OnMessageStatusChangedListener {
            override fun messageReceived(message: Message) {
                messages.add(0, MessageStatus(message).received())
                //почистить очередь больше 100
                if (messages.size > MESSAGE_LIMIT) {
                    messages.subList(MESSAGE_LIMIT, messages.size).clear()
                }
            }

            override fun messageAnswered(message: Message) {
                for (messageStatus in messages) {
                    messageStatus.message?.let {
                        if (it.message_id == message.message_id) {
                            messageStatus.answered().message = message
                        }
                    }
                }
            }

            override fun messageError(message: Message, e: Exception) {
                for (messageStatus in messages) {
                    messageStatus.message?.let {
                        if (it.message_id == message.message_id) {
                            messageStatus.error().message = message
                        }
                    }
                }
            }

            override fun messageIgnored(message: Message) {
                for (messageStatus in messages) {
                    messageStatus.message?.let {
                        if (it.message_id == message.message_id) {
                            messageStatus.ignored().message = message
                        }
                    }
                }
            }
        })
    }

    companion object {
        private const val MESSAGE_LIMIT = 500
    }
}